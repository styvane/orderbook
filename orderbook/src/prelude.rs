pub use super::error::Error;
pub use super::integration::runtime;
pub use super::order_book::*;

pub type Result<T> = std::result::Result<T, Error>;
