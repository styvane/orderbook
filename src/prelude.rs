pub use super::configuration::Configuration;
pub use super::error::Error;
pub use super::integration::runtime;
pub use super::integration::summary;
pub use super::integration::summary::SummaryService;
pub use super::order_book::*;

pub type Result<T> = std::result::Result<T, Error>;
