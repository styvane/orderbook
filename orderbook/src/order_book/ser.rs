use serde::ser::{Serialize, SerializeStruct, Serializer};

use super::{BookKind, OrderBook};

impl Serialize for OrderBook {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut state = serializer.serialize_struct("OrderBook", 2)?;
        let books = self
            .books
            .clone()
            .into_sorted_iter()
            .rev()
            .map(|(e, mut b)| {
                b.exchange = Some(e);
                b
            })
            .collect::<Vec<_>>();

        let kind = match self.kind {
            BookKind::Asks => "asks",
            BookKind::Bids => "bids",
        };
        state.serialize_field(kind, &books)?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use crate::prelude::*;
    use rust_decimal_macros::dec;

    #[test]
    fn serde_can_serialize_orderbook() {
        let mut order_book = OrderBook::with_capacity(BookKind::Asks, 2);
        let book = Book::new(dec!(2.1), dec!(0.4));
        order_book.push(Exchange::Bitstamp, book);
        let book = Book::new(dec!(3.1), dec!(0.1));
        order_book.push(Exchange::Binance, book);
        insta::assert_json_snapshot!(&order_book);
    }
}
