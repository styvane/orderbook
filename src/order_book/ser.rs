use serde::ser::{Serialize, SerializeStruct, Serializer};

use super::{BookKind, BookQueue};

impl Serialize for BookQueue {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut state = serializer.serialize_struct("BookQueue", 2)?;
        let books = self
            .books
            .clone()
            .into_sorted_iter()
            .rev()
            .map(|(_, b)| b)
            .collect::<Vec<_>>();

        let kind = match self.kind {
            BookKind::Asks => "asks",
            BookKind::Bids => "bids",
        };
        state.serialize_field(kind, &books)?;
        state.end()
    }
}

#[cfg(test)]
mod tests {
    use crate::prelude::*;

    #[test]
    fn serde_can_serialize_orderbook() {
        let mut order_book = BookQueue::with_capacity(BookKind::Asks, 2);
        let book = Book::new("2.1", "0.4", "bitstamp");
        order_book.push(Exchange::Bitstamp, book);
        let book = Book::new("3.1", "0.1", "binance");
        order_book.push(Exchange::Binance, book);
        insta::assert_json_snapshot!(&order_book);
    }
}
