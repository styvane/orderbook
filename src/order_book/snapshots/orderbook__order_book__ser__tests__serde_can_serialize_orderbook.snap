---
source: orderbook/src/order_book/ser.rs
expression: "&order_book"
---
{
  "asks": [
    {
      "exchange": "binance",
      "price": "3.1",
      "amount": "0.1"
    },
    {
      "exchange": "bitstamp",
      "price": "2.1",
      "amount": "0.4"
    }
  ]
}
